File: src/java/version/src/main/java/com/mooltiverse/oss/nyx/version/CompositeStringIdentifier.java
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooltiverse.oss.nyx.version;

import java.util.List;

/**
 * A composite identifier that only accepts {@link StringIdentifier}s as children.
 */
class CompositeStringIdentifier extends CompositeIdentifier {
    /**
     * Serial version UID to comply with {@link java.io.Serializable}
     */
    private static final long serialVersionUID = 1L;

    /**
     * Builds the identifier with the given nested identifiers and separator.
     *
     * @param separator the separator used for the string representation to separate children values
     * @param children the children of this composite identifier. It can't be {@code null} or contain {@code null} values
     *
     * @throws NullPointerException if the given list of children is {@code null} or contains {@code null} values
     * @throws IllegalArgumentException if the given list of children contains illegal values
     */
    protected CompositeStringIdentifier(char separator, StringIdentifier... children) {
        super(separator, children);
    }

    /**
     * Builds the identifier with the given nested identifiers and the default separator.
     *
     * @param children the children of this composite identifier. It can't be {@code null} or contain {@code null} values
     *
     * @throws NullPointerException if the given list of children is {@code null} or contains {@code null} values
     * @throws IllegalArgumentException if the given list of children contains illegal values
     */
    protected CompositeStringIdentifier(StringIdentifier... children) {
        super(children);
    }

    /**
     * Builds the identifier with the given nested identifiers and separator.
     *
     * @param separator the separator used for the string representation to separate children values
     * @param children the children of this composite identifier. It can't be {@code null} or contain {@code null} values
     *
     * @throws NullPointerException if the given list of children is {@code null} or contains {@code null} values
     * @throws IllegalArgumentException if the given list of children contains illegal values
     */
    protected CompositeStringIdentifier(char separator, List<StringIdentifier> children) {
        super(separator, children);
    }

    /**
     * Builds the identifier with the given nested identifiers and the default separator.
     *
     * @param children the children of this composite identifier. It can't be {@code null} or contain {@code null} values
     *
     * @throws NullPointerException if the given list of children is {@code null} or contains {@code null} values
     * @throws IllegalArgumentException if the given list of children contains illegal values
     */
    protected CompositeStringIdentifier(List<StringIdentifier> children) {
        super(children);
    }

    /**
     * Returns an identifier instance representing the specified String value. The string is split using the
     * default separator.
     *
     * @param s the string to parse
     *
     * @return the new identifier instance representing the given string.
     *
     * @throws NullPointerException if the given string is {@code null}
     * @throws IllegalArgumentException if the given string contains illegal characters
     */
    static CompositeStringIdentifier valueOf(String s) {
        return valueOf(s, DEFAULT_SEPARATOR);
    }

    /**
     * Returns an identifier instance representing the specified String value.
     *
     * @param s the string to parse
     * @param separator the separator used for the string representation to separate children values
     *
     * @return the new identifier instance representing the given string.
     *
     * @throws NullPointerException if the given string is {@code null}
     * @throws IllegalArgumentException if the given string contains illegal characters
     */
    static CompositeStringIdentifier valueOf(String s, char separator) {
        return new CompositeStringIdentifier(separator, Parser.toStringIdentifiers(s, separator));
    }

    /**
     * Returns the string at the given position (the first being at position 0)
     *
     * @param i the position to retrieve the value at (starting from 0)
     *
     * @return the string at the given position
     *
     * @throws IndexOutOfBoundsException if the given index is out of range
     */
    public String get(int i) {
        return String.class.cast(children.get(i).getValue());
    }
}

File: src/go/version/composite_string_identifier.go
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package version

import (
	"fmt"     // https://pkg.go.dev/fmt
	"reflect" // https://pkg.go.dev/reflect
)

/*
A composite identifier that only accepts StringIdentifierss as children.
*/
type compositeStringIdentifier struct {
	// Extend compositeIdentifier by composition
	compositeIdentifier

	// The nested identifiers. This is redundant with the member from the nested compositeIdentifier
	// but it's typed so we can save methods with (up/down) casting.
	children []stringIdentifier
}

/*
Builds the identifier with the given nested identifiers and separator.
*/
func newCompositeStringIdentifierWithSeparator(separator string, children ...stringIdentifier) (compositeStringIdentifier, error) {
	identifier := compositeStringIdentifier{}

	identifiers, err := castSliceOfStringIdentifiersToSliceOfIdentifiers(children)
	if err != nil {
		return compositeStringIdentifier{}, fmt.Errorf("unexpected error when downcasting identifiers: %w", err)
	}

	identifier.compositeIdentifier.separator = separator
	identifier.compositeIdentifier.children = identifiers
	identifier.children = children

	return identifier, nil
}

/*
Builds the identifier with the given nested identifiers and the default separator.
*/
func newCompositeStringIdentifier(children ...stringIdentifier) (compositeStringIdentifier, error) {
	identifier := compositeStringIdentifier{}

	identifiers, err := castSliceOfStringIdentifiersToSliceOfIdentifiers(children)
	if err != nil {
		return compositeStringIdentifier{}, fmt.Errorf("unexpected error when downcasting identifiers: %w", err)
	}

	identifier.compositeIdentifier.separator = DEFAULT_SEPARATOR
	identifier.compositeIdentifier.children = identifiers
	identifier.children = children

	return identifier, nil
}

/*
Transforms the given slice of elements in a slice with the same elements of type
identifier.
*/
func castSliceOfStringIdentifiersToSliceOfIdentifiers(ids []stringIdentifier) ([]identifier, error) {
	res := make([]identifier, len(ids))
	for i, item := range ids {
		res[i] = item
	}
	return res, nil
}

/*
Returns the string at the given position (the first being at position 0)
*/
func (csi compositeStringIdentifier) get(i uint) string {
	return csi.children[i].getStringValue()
}

/*
Tests the given object for equality against this instance.
*/
func (csi compositeStringIdentifier) equals(obj interface{}) bool {
	if obj == nil {
		return false
	}
	otherIdentifier, ok := obj.(compositeStringIdentifier)
	if !ok {
		return false
	}
	if csi.compositeIdentifier.separator != otherIdentifier.compositeIdentifier.separator {
		return false
	}
	return reflect.DeepEqual(csi.children, otherIdentifier.children)
}
